### ISSUE TITLE
RF01 and redshift/dbt source function

### ISSUE THREAD

**Issue by kochalex** [sqlfluff/sqlfluff#6583]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

sqlfluff 3.3.0 finds a RF01 when using a dbt source jinja function without an alias inline with a query -  e.g. just on a join line and not in a CTE (which is very common in dbt). This behavior did not error in 3.2.5.

### Expected Behaviour

Aliases are not required when using the source function as part of query.

### Observed Behaviour

This SQL has a RF01 error on the first join only

```sql
with cte_table as (select * from {{ source('public', 'events') }})

select count(*) from cte_table
inner join {{ source('public', 'events') }}
    on events.user_id = cte_table.user_id
inner join {{ source('public', 'events') }} as events_aliased
    on events_aliased.user_id = cte_table.user_id

```

```
sqlfluff lint models/test.sql
=== [dbt templater] Sorting Nodes...
=== [dbt templater] Compiling dbt project...
=== [dbt templater] Project Compiled.
== [models/test.sql] FAIL
L:   5 | P:   8 | RF01 | Reference 'events.user_id' refers to table/view not
                       | found in the FROM clause or found in ancestor statement.
                       | [references.from]
All Finished 📜 🎉!
```

### How to reproduce

1. Create a dbt/sqlfluff/redshift project with at least one source defined (here `public.events`)
2. Create the following model
```sql
with cte_table as (select * from {{ source('public', 'events') }})

select count(*) from cte_table
inner join {{ source('public', 'events') }}
    on events.user_id = cte_table.user_id
inner join {{ source('public', 'events') }} as events_aliased
    on events_aliased.user_id = cte_table.user_id

```
3. sqlfluff config:
```
[sqlfluff]
dialect = redshift
templater = dbt
```

4. Run `sqlfluff lint models/test.sql`

### Dialect

redshift

### Version

sqlfluff version output: `sqlfluff, version 3.3.0`
sqlfluff-templater-dbt: 3.3.0
python: 3.12.7
dbt: 1.9.1

### Configuration

```
[sqlfluff]
dialect = redshift
templater = dbt
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
